File: backend-fastapi/api/routes/hello_world.py
from fastapi import APIRouter, Depends

from api.middlewares import get_current_admin, get_current_user
from background.publishers import send_newsletter
from models.schemas import UserReadSchema

hello_world_router = APIRouter(prefix="/api/hello-world", tags=["Hello World"])


@hello_world_router.get(
    "",
    summary="Hello World Endpoint",
    responses={
        200: {
            "content": {
                "application/json": {"example": {"message": "Hello World!"}}
            },
        }
    },
)
async def hello():
    send_newsletter("Slim", "mslimbeji@gmail.com")
    return dict(message="Hello World!")


@hello_world_router.get(
    "/user",
    summary="Hello World Endpoint for authenticated users",
    responses={
        200: {
            "content": {
                "application/json": {"example": {"message": "Hello Slim!"}}
            },
        }
    },
)
async def hello_user(user: UserReadSchema = Depends(get_current_user)):
    return dict(message=f"Hello {user.name}!")


@hello_world_router.get(
    "/admin",
    summary="Hello World Endpoint for admins only",
    responses={
        200: {
            "content": {
                "application/json": {
                    "example": {"message": "Hello Admin Slim!"}
                }
            },
        }
    },
)
async def hello_admin(user: UserReadSchema = Depends(get_current_admin)):
    return dict(message=f"Hello Admin {user.name}!")

File: backend-express/src/api/routes/helloWorld.ts
import { Request, Response, Router } from "express";

import { sendNewsletter } from "@/background/publishers";
import { zod } from "@/lib/zod";
import { UserRead } from "@/models/schemas";

import { swaggerRegistery } from "../docs";
import { Admin, Authenticated } from "../middlewares";

export const helloWorldRouter = Router();

// Hello World Endpoint
async function hello(req: Request, res: Response) {
    sendNewsletter("Slim", "mslimbeji@gmail.com");
    res.status(200).json({ message: "Hello World!" });
}

helloWorldRouter.get("/", hello);

swaggerRegistery.registerPath({
    method: "get",
    path: "/hello-world",
    responses: {
        200: {
            description: "A simple response with message field",
            content: {
                "application/json": {
                    schema: zod.object({
                        message: zod.string().openapi({
                            example: "Hello World",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Hello World"],
    summary: "Hello World Endpoint",
});

// Authenticated Hello World
async function helloUser(req: Request, res: Response) {
    const user = req.currentUser as UserRead;
    res.status(200).json({ message: `Hello ${user.name}!` });
}

helloWorldRouter.get("/user", Authenticated, helloUser);

swaggerRegistery.registerPath({
    method: "get",
    path: "/hello-world/user",
    responses: {
        200: {
            description: "A simple response with message field",
            content: {
                "application/json": {
                    schema: zod.object({
                        message: zod.string().openapi({
                            example: "Hello Slim",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Hello World"],
    summary: "Hello World Endpoint for authenticated users",
    security: [{ OAuth2Password: [] }],
});

// Hello World for admins
async function helloAdmin(req: Request, res: Response) {
    const user = req.currentUser as UserRead;
    res.status(200).json({
        message: `Hello Admin ${user.name}!`,
    });
}

helloWorldRouter.get("/admin", Admin, helloAdmin);

swaggerRegistery.registerPath({
    method: "get",
    path: "/hello-world/admin",
    responses: {
        200: {
            description: "A simple response with message field",
            content: {
                "application/json": {
                    schema: zod.object({
                        message: zod.string().openapi({
                            example: "Hello Admin Slim",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Hello World"],
    summary: "Hello World Endpoint for admins only",
    security: [{ OAuth2Password: [] }],
});
